### ISSUE TITLE
Assertion failed 'rhs->OperGet() == GT_LONG'

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#18986]
This is a recent regression.

Every Ubuntu arm32 JitStress=2 with JitStressRegs failed with:
```
FAILED   - JIT/Methodical/fp/exgen/10w250d_cs_ro/10w250d_cs_ro.sh
               BEGIN EXECUTION
               /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_jitstress2_jitstressregs1_tst/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun 10w250d_cs_ro.exe
               
               Assert failure(PID 23405 [0x00005b6d], Thread: 23405 [0x5b6d]): Assertion failed 'rhs->OperGet() == GT_LONG' in 'testout1:Func_0_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1_1():double' (IL size 788)
               
                   File: /mnt/j/workspace/dotnet_coreclr/master/arm_cross_checked_ubuntu/src/jit/decomposelongs.cpp Line: 433
                   Image: /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_jitstress2_jitstressregs1_tst/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun
               
               ./10w250d_cs_ro.sh: line 244: 23405 Aborted                 $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED
...
FAILED   - JIT/Methodical/fp/exgen/10w5d_cs_ro/10w5d_cs_ro.sh
               BEGIN EXECUTION
               /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_jitstress2_jitstressregs1_tst/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun 10w5d_cs_ro.exe
               Func_0_6_5_6_3: != true
               Func_0_6_5_4_6: <= true
               Func_0_6_5_4_5: > true
               Func_0_6_5_4_3: == false
               Func_0_6_4_6_2: != true
               Func_0_6_4_5_4: < true
               Func_0_6_4_5_1: < true
               Func_0_6_4_4: <= true
               Func_0_6_3_6_5: == false
               Func_0_6_3_4_4: <= true
               Func_0_6_3_4_3: < true
               Func_0_6_3_3: < true
               Func_0_6_3_2_3: >= true
               Func_0_6_3_1_5: < true
               Func_0_6_3_1: <= false
               Func_0_6_3: != true
               Func_0_6_2_5: != true
               Func_0_6_2_1: > false
               Func_0_6_1_6_4: < false
               Func_0_6_1_6_3: > false
               Func_0_6_1_5_4: > true
               
               Assert failure(PID 23525 [0x00005be5], Thread: 23525 [0x5be5]): Assertion failed 'rhs->OperGet() == GT_LONG' in 'testout1:Func_0_6_1_3_3():double' (IL size 157)
               
                   File: /mnt/j/workspace/dotnet_coreclr/master/arm_cross_checked_ubuntu/src/jit/decomposelongs.cpp Line: 433
                   Image: /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_jitstress2_jitstressregs1_tst/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun
               
               ./10w5d_cs_ro.sh: line 244: 23525 Aborted                 $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED
```

e.g., https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_jitstress2_jitstressregs1_tst/15/consoleText.

There is currently no recent run without JitStressRegs to see if it fails with just JitStress=2 (https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_jitstress2_flow/)

@dotnet/arm32-contrib @dotnet/jit-contrib

**Comment by BruceForstall** [comment]
It happens with just JitStress=2, as seen in the Windows arm32 runs:

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_windows_nt_jitstress2_flow/39/

**Comment by AndyAyersMS** [comment]
Also seeing this on x86
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_jitstress2_prtest/92/testReport/junit/JIT_Methodical/_fp_exgen_10w5d_cs_ro_10w5d_cs_ro_/_fp_exgen_10w5d_cs_ro_10w5d_cs_ro_cmd/

**Comment by sandreenko** [comment]
It stopped failing after #19359 (Change CoreLib native image to be R2R by default on all platforms)  was reverted by #19501.

Link #19502.

**Comment by sandreenko** [comment]
Ok, it is another issue with struct promotion of `Low64`
```
        private ulong Low64
        {
#if BIGENDIAN
            get { return ((ulong)umid << 32) | ulo; }
            set { umid = (uint)(value >> 32); ulo = (uint)value; }
#else
            get => ulomidLE;
            set => ulomidLE = value;
#endif
        }
```

The difference between this isse and #18988 is that in this issue the tree has an assignment:              
```
[000601] ------------              *  STMT      void  (IL   ???...  ???)
[000854] ------------              |  /--*  FIELD     long   ulomidLE <- long is the right type.
[000852] L-----------              |  |  \--*  ADDR      byref 
[000853] ------------              |  |     \--*  LCL_VAR   struct V16 tmp15        
[000600] -AC---------              \--*  ASG       long  
[000599] D------N----                 \--*  LCL_VAR   long   V30 tmp29 
```
that we corrupt with "Replacing the field in promoted struct with local var V46":
```
[000601] ------------              *  STMT      void  (IL   ???...  ???)
[000854] ------------              |  /--*  LCL_VAR   int   (AX) V46 tmp45   <- BAD type.   
[000600] -AC---------              \--*  ASG       long  
[000599] D------N----                 \--*  LCL_VAR   long   V30 tmp29 
```
we should check in `fgMorphStructField` that types match.
cc @mikedn

**Comment by sandreenko** [comment]
> Looks like. But this might be a bug in struct promotion itself, I didn't know it promotes structs with overlapping fields. I'll take a look.

There are no overlapping fields in this example.

> We shouldn't be perform struct promotion on Structs with explicit layout.

Why should not we? Structs with explicit layout are allocated on stack, so it should guarantee the layout of the fields. Could we have a ReadWrite or WriteRead conflict in this case?

**Comment by mikedn** [comment]
> There are no overlapping fields in this example.

The Low64 property you are showing comes from this struct:

```C#
        [StructLayout(LayoutKind.Explicit)]
        private struct DecCalc
        {
            // NOTE: Do not change the offsets of these fields. This structure must have the same layout as Decimal.
            [FieldOffset(0)]
            private uint uflags;
            [FieldOffset(4)]
            private uint uhi;
            [FieldOffset(8)]
            private uint ulo;
            [FieldOffset(12)]
            private uint umid;

            /// <summary>
            /// The low and mid fields combined in little-endian order
            /// </summary>
            [FieldOffset(8)]
            private ulong ulomidLE;
```

**Comment by briansull** [comment]
>> We shouldn't be perform struct promotion on Structs with explicit layout.

> Why should not we? Structs with explicit layout are allocated on stack

It is not a correctness issue, but I believe it is used to simplfy our implementatiion.  
Rather that trying to determined if any overlapping occurs with a given struct's explicit layout 
instead we just return false when wes ee the struct has an explict layout.

**Comment by sandreenko** [comment]
> > There are no overlapping fields in this example.

> The Low64 property you are showing comes from this struct:

Yes, but there are two structs: `Decimal` and `DecCalc`, `DecCalc` has `StructLayout(LayoutKind.Explicit)]` and overlapping fields. I do not think that it makes `DecCalc` a struct with overlapping fields, does it?

and `Decimal` fields are declared as:
```
        private readonly int flags;
        private readonly int hi;
        private readonly int lo;
        private readonly int mid;
```
so they do not overlap.

**Comment by sandreenko** [comment]
> We need to pay careful attention to these "struct casts".
> If an otherwise promotable struct instance A is cast to an some other struct instance B, then A is in general no longer promotable, since its fields now interfere with the fields of B. Basically with the cast the user has created an ad-hoc union type.
> This is symmetric and happens whether or not B is promotable... so if both A and B seemed to be promotable before seeing the cast, neither one is afterwards.

It looks too conservative, I think even VM can't give us this information, because we always can load a library  that adds new struct casts when we have already compiled other methods using struct promotions.

**Comment by AndyAyersMS** [comment]
We generally don't need to worry about casts that aren't visible in the current method.

Implicit by-ref struct args, by-value struct args, and local structs are initially presumed to have unaliased fields. So we just need to look for evidence to the contrary within the method.

Evidence would be overlapping fields, addresses escaping, wider than field accesses, and casts of a byref for the struct to a different struct type.

**Comment by CarolEidt** [comment]
I believe that @AndyAyersMS is correct in his assessment. Before `Unsafe.As` such casts were not achievable from C# (I don't think), so no surprise that this issue didn't arise previously.
